Decoder

derzhavin3016 opened the pull request:
- Renamed Reg to RegVal
- Added Addr & RegId types
- Added kRegNum constant
- Added enum w/ operations

closing #8

derzhavin3016 committed:
[common] Added useful types + constants

- Renamed Reg to RegVal
- Added Addr & RegId types
- Added kRegNum constant
- Added enum w/ operations

derzhavin3016 committed:
[state] Added throw in case of set(x0)

derzhavin3016 committed:
[decoder] Added Instruction class + getBits

derzhavin3016 committed:
[decoder] Added simple getBits() test

derzhavin3016 committed:
[decoder] Fix decoder tests & include dir for test

derzhavin3016 committed:
[decoder] Added sign extend function

derzhavin3016 committed:
[decoder] Simplified signedExtend algo

derzhavin3016 committed:
[infra] Fix threads in build CI & disabled format

Relates to #18

derzhavin3016 committed:
[decoder] consteval sizeofBits + added setBit()

derzhavin3016 committed:
[decoder] Added riscv opcodes as submodule

derzhavin3016 committed:
[decoder] Added signExtend to full Word

derzhavin3016 committed:
[decoder] Codegen first steps

derzhavin3016 committed:
[decoder] Changed signedBits algo to a safer one

- Implemented algo which operates only on unsigned numbers

derzhavin3016 committed:
[docs] Added spec

derzhavin3016 committed:
[decoder] Finished codegen impl

Tako-San commented on include/decoder/decoder.hh:
```diff
@@ -3,12 +3,55 @@
 
 #include "common/common.hh"
 #include "common/inst.hh"
+#include <type_traits>
```
nit: Move it up, separately from our include files

derzhavin3016 replied:
Fixed

derzhavin3016 committed:
[decoder] Tests + fixes

- Added processing case newSize == oldSize in signExtend via if constexpr
- Added rv32_i to codegen (SRAI, SRLI, SLLI ...) + shamtw field
- Fixed bug w/ forgotten +1 in signExtend template arg in codegen (it must be size, not index)
- Added tests for all immediate types & additional fields which for now are treated as immediate

derzhavin3016 committed:
[decoder] Moved bit-manipulation funcs to common

derzhavin3016 committed:
[decoder] Added docs comments

Tako-San commented on include/common/common.hh:
```diff
@@ -6,7 +6,58 @@
+  return word & ~mask;
+}
+
+template <std::size_t newSize, std::size_t oldSize> Word signExtend(Word word) {
+  static_assert(newSize >= oldSize, "Trying to sign extend to smaller size");
+  static_assert(oldSize > 0, "Initial size must be non-zero");
+  static_assert(newSize <= sizeofBits<Word>(), "newSize is out of bits range");
+
+  if constexpr (newSize == oldSize)
+    return word;
+
+  Word zeroed = getBits<oldSize - 1, 0>(word);
+  constexpr Word mask = Word(1) << (oldSize - 1);
+  Word res = (zeroed ^ mask) - mask;
+
+  return getBits<newSize - 1, 0>(res);
+}
```
nit: Maybe it is good idea to provide a nice-looking comment with example of work of this algorithm

derzhavin3016 replied:
![изображение](https://user-images.githubusercontent.com/37047102/198723912-8a0474af-a276-42aa-9557-5ccef29c7784.png)

Tako-San replied:
Cool

derzhavin3016 committed:
[decoder, docs] Added doxygen + docstring

- Added doxygen support + documentation + option
- Added docstring to signExtend

Tako-San commented on src/decoder/CMakeLists.txt:
```diff
@@ -1 +1,49 @@
+  COMPONENTS Interpreter
+  REQUIRED)
+
+set(SIM_RISCV_DIR ${CMAKE_CURRENT_SOURCE_DIR}/../../thirdparty/riscv-opcodes)
```
nit: maybe it is better to use CMAKE_SOURCE_DIR

Tako-San reviewed (approved):
LGTM
